### PR TITLE
Deploy UI within fidesops

### LINKED ISSUE #528: [Admin UI] Configure server to host static files
### Is your feature request related to a specific problem?

The Fidesops application server should be able to serve static media.

### Describe the solution you'd like

_A description of what you want to happen._

### Describe alternatives you've considered, if any

_A description of any alternative solutions or features you've considered._

### Additional context

_Add any other context or screenshots about the feature request here._

### PR DESCRIPTION
**Pull request by TheAndrewJackson**
# Purpose
Build and serve the Admin UI within fidesops as a static website.

# Changes
- Update Dockerfile to build website and copy files into the project
- Add `startup` event to serve the website if it's toggled on
- Fix a few miscellaneous issues with serving the frontend from within fidesops
- Refactor `Head` portion of pages into single `FidesHead` component to remove duplication

# Checklist
- [x] Update [`CHANGELOG.md`](https://github.com/ethyca/fidesops/blob/main/CHANGELOG.md) file
  - [x] Merge in main so the most recent `CHANGELOG.md` file is being appended to
  - [x] Add description within the `Unreleased` section in an appropriate category. Add a new category from the list at the top of the file if the needed one isn't already there.
  - [x] Add a link to this PR at the end of the description with the PR number as the text. example: [#1](https://github.com/ethyca/fidesops/pull/1)
- [x] Applicable documentation updated (guides, quickstart, postman collections, tutorial, fidesdemo, [database diagram](https://github.com/ethyca/fidesops/blob/main/docs/fidesops/docs/development/update_erd_diagram.md).
- If docs updated (select one):
  - [x] documentation complete, or draft/outline provided (tag docs-team to complete/review on this branch)
  - [ ] documentation issue created (tag docs-team to complete issue separately)
- [x] Good unit test/integration test coverage
- [ ] This PR contains a DB migration. If checked, the reviewer should confirm with the author that the [down_revision correctly references the previous migration](https://ethyca.github.io/fidesops/development/contributing_details/#alembic-migrations) before merging
- [ ] The `Run Unsafe PR Checks` label has been applied, and checks have passed, if this PR touches any external services

# Ticket

Fixes #528 #529

### CODE REVIEW

File: docs/fidesops/docs/guides/configuration_reference.md [hunk 1]
```diff
@@ -60,8 +60,8 @@ The `fidesops.toml` file should specify the following variables:
 |`MASKING_STRICT` | `FIDESOPS__EXECUTION__MASKING_STRICT` | bool | True | True | If MASKING_STRICT is True, we only use "update" requests to mask data. (For third-party integrations, you should define an `update` endpoint to use.)  If MASKING_STRICT is False, you are allowing fidesops to use any defined DELETE or GDPR DELETE endpoints to remove PII. In this case, you should define `delete` or `data_protection_request` endpoints for your third-party integrations.  Note that setting MASKING_STRICT to False means that data may be deleted beyond the specific data categories that you've configured in your Policy.
 |---|---|---|---|---|---|
 |`ANALYTICS_OPT_OUT` | `FIDESOPS__USER__ANALYTICS_OPT_OUT` | bool | False | False | Opt out of sending anonymous usage data to Ethyca to improve the product experience
-
-
+| Admin UI Variables|---|---|---|---|---|
+|`ENABLED` | `FIDESOPS__ADMIN_UI__ENABLED` | bool | False | True | Toggle whether the `/static` file directory is mounted to serve the Admin UI
```
Comment (conceptualshark) [inline docs/fidesops/docs/guides/configuration_reference.md:64]: Thank you for this! 👍

**Comment by eastandwestwind** [comment]
@TheAndrewJackson I'm getting this build err locally when trying to `make server`, any ideas?


```
Building fidesops
[+] Building 1.2s (6/6) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                   0.0s
 => => transferring dockerfile: 2.08kB                                                                                                                                                                 0.0s
 => [internal] load .dockerignore                                                                                                                                                                      0.0s
 => => transferring context: 208B                                                                                                                                                                      0.0s
 => ERROR [internal] load metadata for docker.io/library/python:3.9.13-slim-buster                                                                                                                     1.1s
 => ERROR [internal] load metadata for docker.io/library/node:16                                                                                                                                       1.1s
 => [auth] library/python:pull token for registry-1.docker.io                                                                                                                                          0.0s
 => [auth] library/node:pull token for registry-1.docker.io                                                                                                                                            0.0s
------
 > [internal] load metadata for docker.io/library/python:3.9.13-slim-buster:
------
------
 > [internal] load metadata for docker.io/library/node:16:
------
failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to authorize: rpc error: code = Unknown desc = failed to fetch oauth token: unexpected status: 500 Internal Server Error
ERROR: Service 'fidesops' failed to build : Build failed
make: *** [compose-build] Error 1
```

**Comment by eastandwestwind** [comment]
Aside from my comment above ^ the code itself looks good. 

Not blocking, but would like to add documentation for how to nav to the admin ui when the `fidesops` server is up and running.

**Comment by TheAndrewJackson** [comment]
> @TheAndrewJackson I'm getting this build err locally when trying to `make server`, any ideas?
> 
> ```
> Building fidesops
> [+] Building 1.2s (6/6) FINISHED
>  => [internal] load build definition from Dockerfile                                                                                                                                                   0.0s
>  => => transferring dockerfile: 2.08kB                                                                                                                                                                 0.0s
>  => [internal] load .dockerignore                                                                                                                                                                      0.0s
>  => => transferring context: 208B                                                                                                                                                                      0.0s
>  => ERROR [internal] load metadata for docker.io/library/python:3.9.13-slim-buster                                                                                                                     1.1s
>  => ERROR [internal] load metadata for docker.io/library/node:16                                                                                                                                       1.1s
>  => [auth] library/python:pull token for registry-1.docker.io                                                                                                                                          0.0s
>  => [auth] library/node:pull token for registry-1.docker.io                                                                                                                                            0.0s
> ------
>  > [internal] load metadata for docker.io/library/python:3.9.13-slim-buster:
> ------
> ------
>  > [internal] load metadata for docker.io/library/node:16:
> ------
> failed to solve with frontend dockerfile.v0: failed to create LLB definition: failed to authorize: rpc error: code = Unknown desc = failed to fetch oauth token: unexpected status: 500 Internal Server Error
> ERROR: Service 'fidesops' failed to build : Build failed
> make: *** [compose-build] Error 1
> ```

I ran into that issue this morning and resolved it by restarting my computer. I'm not completely sure what was wrong 🤷‍♂️

**Comment by TheAndrewJackson** [comment]
> Aside from my comment above ^ the code itself looks good.
> 
> Not blocking, but would like to add documentation for how to nav to the admin ui when the `fidesops` server is up and running.

Good idea. Can we get that in a follow up docs ticket? @eastandwestwind

**Comment by eastandwestwind** [comment]
I was able to resolve the `make server` issues, and got the app running.

Main question is, i'm receiving the placeholder `index.html` instead of the `index.html` supposedly copied over from the build in `Dockerfile`. Is this expected?

**Comment by eastandwestwind** [comment]
I've added a separate issue for docs here- https://github.com/ethyca/fidesops/issues/644

**Comment by TheAndrewJackson** [comment]
> I was able to resolve the `make server` issues, and got the app running.
> 
> Main question is, when I nav to `http://0.0.0.0:8080/static/index.html`, i'm receiving the [placeholder](https://github.com/ethyca/fidesops/pull/625/files#diff-51868caf44d288309a19f7b59360f148adfa01226f70bf528441cef723439ffcR57) `index.html` instead of the `index.html` supposedly copied over from the build in `Dockerfile` [here](https://github.com/ethyca/fidesops/pull/625/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557R59). Is this expected?

Ahh, it's sort of expected. I ran into this a few times while developing this feature. What you're most likely seeing is the `docker-compose.yml` file is removing your UI files. Here is the snippet that is causing the issue:
```yaml
    volumes:
      - type: bind
        source: ./
        target: /fidesops
        read_only: False
      - /fidesops/src/fidesops.egg-info
```

What's happening is the UI is built when the docker image is created initially. Then when `make server` or `docker-compose up` is run it mounts the local `fidesops` directory which doesn't contain the `src/fidesops/build/static` path so it creates the placeholder. If you comment out the snippet above in the `docker-compose.yml` file it should serve the UI like you're expecting as long as the initial docker image has the UI built into it.

**Comment by eastandwestwind** [comment]
Thanks for the clarification @TheAndrewJackson !

**Review by eastandwestwind — APPROVED** [review]

File: src/fidesops/main.py [hunk 2]
```diff
@@ -43,6 +45,21 @@
...
+@app.on_event("startup")
+async def create_webapp_dir_if_not_exists() -> None:
+    """Creates the webapp directory if it doesn't exist."""
+    if config.admin_ui.ENABLED:
+        WEBAPP_DIRECTORY = Path("src/fidesops/build/static")
+        WEBAPP_INDEX = WEBAPP_DIRECTORY / "index.html"
+        if not WEBAPP_INDEX.is_file():
+            WEBAPP_DIRECTORY.mkdir(parents=True, exist_ok=True)
+            with open(WEBAPP_DIRECTORY / "index.html", "w") as index_file:
+                index_file.write("<h1>Privacy is a Human Right!</h1>")
```
Comment (NevilleS) [inline src/fidesops/main.py:57]: I'm curious why is this needed? Is this just to handle a case where the docker build hasn't successfully built the admin UI and exported it, or is it more common than that?

File: src/fidesops/core/config.py [hunk 3]
```diff
@@ -128,6 +137,7 @@ class FidesopsConfig(FidesSettings):
     security: FidesopsSecuritySettings
     execution: ExecutionSettings
     root_user: RootUserSettings
+    admin_ui: AdminUiSettings
```
Comment (NevilleS) [inline src/fidesops/core/config.py:140]: Can you add this to the `log_all_config_values` helper too? Sorry I didn't know a clean way to introspect the settings object to do this automatically

File: src/fidesops/main.py (hunk 2 above)
Comment (TheAndrewJackson) [reply src/fidesops/main.py:57]: I followed the same pattern that the `fides` used did for this update: https://github.com/ethyca/fides/blob/main/src/fidesapi/main.py#L59-L68. I believe that it is so there is always something to display in the event that the admin UI is toggled on. 

It should probably be updated to display something for helpful to help the user know why the full admin ui isn't being displayed.

One common issue that Catherine, Sean, and I ran into while this was developed the the default config in the `docker-compose.yml` mounts a volume of the local `fidesops` files with this snippet
```yaml
    volumes:
      - type: bind
        source: ./
        target: /fidesops
        read_only: False
      - /fidesops/src/fidesops.egg-info
```

Since the admin ui is only built while the docker image is created the ui files are missing from the local `fidesops` code causing the ui files to be missing once `docker-compose up` is ran. 

This is okay because frontend development still uses `npm run dev` from `clients/admin-ui` and backend development won't typically need to use the admin ui from within the `fidesops` webserver. In the event that backend development needs the full admin within the webserver they can run the `npm run prod-export` command from within `clients/admin-ui` and that will build and move the full ui into the correct place within the backend code.

File: src/fidesops/core/config.py (hunk 3 above)
Comment (TheAndrewJackson) [reply src/fidesops/core/config.py:140]: Yes, I'll make an issue for it now and update the code.
